Issue #17951: Improve exception message for basedir path mismatch

vivek-0509 opened the pull request:
Issue #17951:

### Problem
When the `basedir` property uses a different path type than the files being processed (for example, a Unix path like `/mnt/...` on Windows), users encounter an unclear error:

`IllegalArgumentException: 'other' is different type of Path`

This message does not clearly indicate that the root cause is a mismatch in the `basedir` configuration, making troubleshooting difficult.

### Solution
Wrapped the `IllegalArgumentException` thrown by `Path.relativize()` inside `CommonUtil.relativizePath()` with a clearer, more informative error message that:

- Displays both the file path and base directory values involved
- Guides users to check and align their `basedir` configuration with the current OS/path format

**Improved error message example:**

```
Failed to relativize path 'C:...' against base directory '/mnt/...'.
"(configured via 'basedir' property). "
Please ensure the base directory path matches the current OS/path type.
```

### Changes
- **CommonUtil.java:** Added try-catch around `Path.relativize()` to throw the new informative exception
- **CommonUtilTest.java:** Added a unit test with Mockito to verify that the improved exception message is thrown correctly

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:
```diff
@@ -264,7 +266,17 @@ public static String relativizePath(final String baseDirectory, final String pat
+            catch (IllegalArgumentException exception) {
+                throw new IllegalStateException(
+                    String.format(Locale.ROOT,
+                        "Failed to relativize path '%s' against base directory '%s' "
```
Good, but please put message to messages.properties
We localize all messages. There are bunch of examples in code on this. But migration is not complete so there still bad examples of English hard coded messages.

vivek-0509 replied:
Okk will improve this

vivek-0509 replied:
Done

romani reviewed (changes requested):
Nice.

Item to improve

vivek-0509 commented on src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java:
```diff
@@ -602,6 +602,39 @@ public void testLoadSuppressionsUriSyntaxException() throws Exception {
+
+        final IllegalStateException exception = getExpectedThrowable(
+            IllegalStateException.class,
+            () -> CommonUtil.relativizePath(baseDirectory, relativePath)
```
Thanks for the clarification, I will update this

vivek-0509 replied:
Done

vivek-0509 commented:
CommonUtil is in the utils package, which is strictly limited by ImportControl. As a result, it cannot use the localization infrastructure:

-  LocalizedMessage not allowed  
-  Definitions not allowed  
-  MessageFormat not allowed

When attempting to use ResourceBundle directly, Checker Framework flagged violations:

- `@UnknownPropertyKey` (key not statically validated)  
- i18nformat.key.not.found (translation key lookup check)

This would require adding suppressions:

- specifier="i18nformat.key.not.found"  
- specifier="argument" (UnknownPropertyKey)

To avoid import restrictions and suppressions, i moved wrapping  out of CommonUtil:

- CommonUtil.relativizePath() now simply throws IllegalArgumentException  
- Checker performs the wrapping

vivek-0509 commented:
I've added the new general.relativizePath message key to all translation files. I want to be transparent about the approach I took:

Issue encountered:

When I initially added translations using native characters (e.g., Russian Cyrillic, Japanese Kanji), I encountered encoding errors:

`Unsupported characters for the charset 'ISO-8859-1'`
Java  .properties
 files use ISO-8859-1 encoding by default, which doesn't support non-ASCII characters

Solution applied:
I converted all non-ASCII translations to Unicode escape sequences (\uXXXX), which is the standard approach for Java i18n

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/Checker.java:
```diff
@@ -664,4 +664,23 @@ Definitions.CHECKSTYLE_BUNDLE, getClass(),
+        try {
+            return CommonUtil.relativizePath(basedir, fileName);
+        }
+        catch (IllegalArgumentException exception) {
```
Catch "Exception" here to be sure we wrap any kind of exception with details

vivek-0509 replied:
done

romani reviewed (changes requested):
Items
